File: tp1/sources/calcul_pi_mpi.py
#!/usr/bin/env python3
"""
TP1 - Section 2.2 : Calcul approché de π - Version mpi4py

Parallélisation en mémoire distribuée avec mpi4py

Installation : pip install mpi4py numpy
Exécution    : mpirun -np 4 python3 calcul_pi_mpi.py [nombre_de_points]
"""

import numpy as np
from mpi4py import MPI
import sys
import time

def main():
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    nbp = comm.Get_size()
    
    # Nombre de points par défaut
    nbSamples = 100_000_000  # 10^8
    
    if len(sys.argv) > 1:
        nbSamples = int(sys.argv[1])
    
    # Chaque processus traite sa portion
    samples_per_proc = nbSamples // nbp
    my_start = rank * samples_per_proc
    my_end = nbSamples if rank == nbp - 1 else my_start + samples_per_proc
    my_samples = my_end - my_start
    
    # Graine unique pour chaque processus (modulo 2^32 pour éviter l'overflow)
    seed = (int(time.time()) + rank * 12345) % (2**32)
    np.random.seed(seed)
    
    # Synchronisation avant le début de la mesure
    comm.Barrier()
    start_time = MPI.Wtime()
    
    # Génération des points et comptage
    x = np.random.uniform(-1.0, 1.0, my_samples)
    y = np.random.uniform(-1.0, 1.0, my_samples)
    local_darts = np.sum(x*x + y*y <= 1.0)
    
    # Réduction : somme de tous les compteurs locaux
    total_darts = comm.reduce(local_darts, op=MPI.SUM, root=0)
    
    # Synchronisation à la fin
    comm.Barrier()
    end_time = MPI.Wtime()
    elapsed = end_time - start_time
    
    # Le processus 0 affiche le résultat
    if rank == 0:
        pi = 4.0 * total_darts / nbSamples
        error = abs(pi - np.pi) / np.pi * 100.0
        
        print("=== Calcul de π - Version mpi4py ===")
        print(f"Nombre de points     : {nbSamples}")
        print(f"Nombre de processus  : {nbp}")
        print(f"π calculé  : {pi:.10f}")
        print(f"π réel     : {np.pi:.10f}")
        print(f"Erreur     : {error:.6f} %")
        print(f"Temps      : {elapsed:.6f} secondes")

if __name__ == "__main__":
    main()

File: tp1/sources/calcul_pi_mpi.cpp
/**
 * TP1 - Section 2.2 : Calcul approché de π - Version MPI
 * 
 * Parallélisation en mémoire distribuée avec MPI
 * Chaque processus génère sa portion de points, puis on utilise MPI_Reduce
 * pour sommer les résultats.
 * 
 * Compilation : mpic++ -O2 -o calcul_pi_mpi.exe calcul_pi_mpi.cpp
 * Exécution   : mpirun -np 4 ./calcul_pi_mpi.exe [nombre_de_points]
 */

#include <iostream>
#include <random>
#include <chrono>
#include <cstdlib>
#include <iomanip>
#include <cmath>
#include <mpi.h>

int main(int argc, char* argv[])
{
    MPI_Init(&argc, &argv);
    
    int rank, nbp;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &nbp);
    
    unsigned long nbSamples = 100000000;  // 10^8 par défaut
    
    if (argc > 1) {
        nbSamples = std::atol(argv[1]);
    }
    
    // Chaque processus traite sa portion
    unsigned long samplesPerProc = nbSamples / nbp;
    unsigned long myStart = rank * samplesPerProc;
    unsigned long myEnd = (rank == nbp - 1) ? nbSamples : myStart + samplesPerProc;
    unsigned long mySamples = myEnd - myStart;
    
    // Graine unique pour chaque processus
    unsigned int seed = std::chrono::system_clock::now().time_since_epoch().count() + rank * 12345;
    std::default_random_engine generator(seed);
    std::uniform_real_distribution<double> distribution(-1.0, 1.0);
    
    // Synchronisation avant le début de la mesure
    MPI_Barrier(MPI_COMM_WORLD);
    double start_time = MPI_Wtime();
    
    // Comptage local des points dans le cercle
    unsigned long localDarts = 0;
    for (unsigned long sample = 0; sample < mySamples; ++sample) {
        double x = distribution(generator);
        double y = distribution(generator);
        if (x * x + y * y <= 1.0) {
            localDarts++;
        }
    }
    
    // Réduction : somme de tous les compteurs locaux
    unsigned long totalDarts = 0;
    MPI_Reduce(&localDarts, &totalDarts, 1, MPI_UNSIGNED_LONG, MPI_SUM, 0, MPI_COMM_WORLD);
    
    // Synchronisation à la fin
    MPI_Barrier(MPI_COMM_WORLD);
    double end_time = MPI_Wtime();
    double elapsed = end_time - start_time;
    
    // Le processus 0 affiche le résultat
    if (rank == 0) {
        double pi = 4.0 * static_cast<double>(totalDarts) / static_cast<double>(nbSamples);
        double error = std::abs(pi - M_PI) / M_PI * 100.0;
        
        std::cout << "=== Calcul de π - Version MPI ===" << std::endl;
        std::cout << "Nombre de points     : " << nbSamples << std::endl;
        std::cout << "Nombre de processus  : " << nbp << std::endl;
        std::cout << std::fixed << std::setprecision(10);
        std::cout << "π calculé  : " << pi << std::endl;
        std::cout << "π réel     : " << M_PI << std::endl;
        std::cout << "Erreur     : " << std::setprecision(6) << error << " %" << std::endl;
        std::cout << "Temps      : " << elapsed << " secondes" << std::endl;
    }
    
    MPI_Finalize();
    return 0;
}
